### ISSUE TITLE
mod_sftp.c: error initializing session: Permission denied

### ISSUE THREAD

**Issue by Dead2** [proftpd/proftpd#160]
Using Proftpd 1.3.6 RC1, on Scientific Linux 6.1 (RHEL6 clone) I am unable to get sftp to work.
```
mod_sftp.c: error initializing session: Permission denied
```

SELinux is disabled with kernel parameter, so that is not the cause.

It says Permission denied, but it does not say what it is attempting to do, thus I am having a hard time figuring this out.

Googling this message, I found the following unresolved report:
https://forums.proftpd.org/smf/index.php?topic=11779.0

Debug log from 
```
# proftpd -4n -d 20 -c /etc/proftpd/proftpd.conf
16:18:05,998 r99.no proftpd[9332]: using TCP receive buffer size of 87380 bytes
16:18:05,998 r99.no proftpd[9332]: using TCP send buffer size of 16384 bytes
16:18:05,998 r99.no proftpd[9332]: testing Unix domain socket using S_ISFIFO
16:18:05,998 r99.no proftpd[9332]: testing Unix domain socket using S_ISSOCK
16:18:05,998 r99.no proftpd[9332]: using S_ISSOCK macro for Unix domain socket detection
16:18:05,999 r99.no proftpd[9332]: using 'UTF-8' as local charset for UTF-8 conversion
16:18:05,999 r99.no proftpd[9332]: retrieved UID 99 for user 'nobody'
16:18:06,000 r99.no proftpd[9332]: retrieved GID 99 for group 'nobody'
16:18:06,000 r99.no proftpd[9332]: mod_dso/0.5: loading 'mod_sftp.c'
16:18:06,000 r99.no proftpd[9332]: mod_dso/0.5: loaded module 'mod_sftp' (from '/usr/libexec/proftpd/mod_sftp.so', last modified on Fri Sep 18 09:35:21 2015)
16:18:06,000 r99.no proftpd[9332]: mod_sftp/1.0.0: using OpenSSL 1.0.1e-fips 11 Feb 2013
16:18:06,000 r99.no proftpd[9332]: <IfModule>: using 'mod_sftp.c' section at line 17
16:18:06,000 r99.no proftpd[9332]: ROOT PRIVS at mod_sftp.c:1107
16:18:06,000 r99.no proftpd[9332]: RELINQUISH PRIVS at mod_sftp.c:1110
16:18:06,000 r99.no proftpd[9332]: mod_sftp/1.0.0: unable to use '/etc/ssh/ssh_host_ecdsa_key' as host key, as it is group- or world-accessible
16:18:06,000 r99.no proftpd[9332]: ROOT PRIVS at mod_sftp.c:1107
16:18:06,000 r99.no proftpd[9332]: RELINQUISH PRIVS at mod_sftp.c:1110
16:18:06,000 r99.no proftpd[9332]: mod_sftp/1.0.0: unable to use '/etc/ssh/ssh_host_rsa_key' as host key, as it is group- or world-accessible
16:18:06,017 r99.no proftpd[9332]: UseReverseDNS off, returning IP address instead of DNS name
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: 
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: Config for SFTP server:
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: ServerIdent
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: DefaultServer
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: DefaultRoot
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: AuthPAMConfig
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: AuthOrder
16:18:06,017 r99.no proftpd[9332] 185.91.65.10: UserID
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: UserName
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: GroupID
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: GroupName
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: SFTPEngine
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: SFTPOptions
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: SFTPHostKey
16:18:06,018 r99.no proftpd[9332] 185.91.65.10: SFTPHostKey
16:18:06,020 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at keys.c:599
16:18:06,020 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at keys.c:602
16:18:06,020 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at keys.c:599
16:18:06,020 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at keys.c:602
16:18:06,021 r99.no proftpd[9332] 185.91.65.10: mod_lang/1.1: binding to text domain 'proftpd' using locale path '/usr/share/locale'
16:18:06,021 r99.no proftpd[9332] 185.91.65.10: mod_lang/1.1: using locale files in '/usr/share/locale'
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at mod_ctrls.c:1199
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at mod_ctrls.c:1201
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at mod_rlimit.c:541
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at mod_rlimit.c:544
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: set core resource limits for daemon
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: retrieved group ID: 99
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: setting group ID: 99
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: SETUP PRIVS at main.c:2436
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at main.c:1743
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: opening scoreboard '/var/run/proftpd/proftpd.scoreboard'
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at main.c:1770
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at inet.c:385
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at inet.c:459
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: ProFTPD 1.3.6rc1 (devel) (built Fri Sep 18 2015 09:34:44 CEST) standalone mode STARTUP
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: ROOT PRIVS at pidfile.c:58
16:18:06,023 r99.no proftpd[9332] 185.91.65.10: RELINQUISH PRIVS at pidfile.c:61
16:18:07,877 r99.no proftpd[9333] 185.91.65.10: ROOT PRIVS at main.c:1177
16:18:07,877 r99.no proftpd[9333] 185.91.65.10: RELINQUISH PRIVS at main.c:1181
16:18:07,877 r99.no proftpd[9333] 185.91.65.10: no matching vhost found for 10.0.10.20#1022, using 'SFTP server' listening on wildcard address
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): ROOT PRIVS at main.c:986
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): SETUP PRIVS at main.c:991
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): session requested from client in unknown class
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): performing module session initializations
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): ROOT PRIVS at keys.c:1969
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): RELINQUISH PRIVS at keys.c:1976
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): ROOT PRIVS at keys.c:1969
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): RELINQUISH PRIVS at keys.c:1976
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): mod_sftp.c: error initializing session: Permission denied
16:18:07,877 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): mod_sftp/1.0.0: scrubbing 2 passphrases from memory
16:18:07,878 r99.no proftpd[9333] 185.91.65.10 (77.40.194.196[77.40.194.196]): FTP session closed.
```

Config used:
```
ServerName                      "SFTP server"
ServerAdmin                     root@localhost
ServerIdent                     off
DefaultServer                   on

DefaultRoot                     ~
UseReverseDNS                   off

AuthPAMConfig                   proftpd
AuthOrder                       mod_auth_pam.c* mod_auth_unix.c

User                            nobody
Group                           nobody

LoadModule mod_sftp.c

<IfModule mod_sftp.c>
    Port 1022
    SFTPEngine on

    SFTPOptions InsecureHostKeyPerms
    SFTPHostKey /etc/ssh/ssh_host_ecdsa_key
    SFTPHostKey /etc/ssh/ssh_host_rsa_key
</IfModule>
```

**Comment by Dead2** [comment]
It seems to be related to host keys, because if I copy the keys to /etc/proftpd/ and chmod 400, then it works.

Is the InsecureHostKeyPerms workaround somehow not doing its thing?

```
# ls -la /etc/ssh
total 284
drwxr-xr-x  2 root root       4096 Sep  4 14:40 .
drwxr-xr-x 77 root root       4096 Sep 25 14:11 ..
-rw-r--r--  1 root root     242153 May 12 19:11 moduli
-rw-r--r--  1 root root       2208 May 12 19:11 ssh_config
-rw-------  1 root root       4361 May 12 19:11 sshd_config
-rw-r-----  1 root ssh_keys    227 Sep  3 16:19 ssh_host_ecdsa_key
-rw-r--r--  1 root root        162 Sep  3 16:19 ssh_host_ecdsa_key.pub
-rw-r-----  1 root ssh_keys    387 Sep  3 16:19 ssh_host_ed25519_key
-rw-r--r--  1 root root         82 Sep  3 16:19 ssh_host_ed25519_key.pub
-rw-r-----  1 root ssh_keys   1679 Sep  3 16:19 ssh_host_rsa_key
-rw-r--r--  1 root root        382 Sep  3 16:19 ssh_host_rsa_key.pub
```

**Comment by Castaglia** [comment]
You might try adding a `SFTPLog` directive, and checking what that log file says.

**Comment by Dead2** [comment]
I added "SFTPLog /var/log/proftpd/sftp.log" to the config file (above the SFTPOptions line), and no log file was created.

The debug output also did not change except it mentions "SFTPLog", in other words I cannot see any permission failures there. Proftpd does create ban.log if I enable that module, so that should not be a problem.

Btw, shouldn't that config printout during debug contain the values of the config parameters, not just what parameters were present?

**Comment by Dead2** [comment]
This was caused by kernel error in our locally patched kernel, ignore.

**Comment by Castaglia** [comment]
Oh, how very interesting.  Thanks for following up on this!  I am now curious just what the kernel error/problem was, that manifested this way in proftpd?

**Comment by Dead2** [comment]
Exactly what part of the code caused the problem is unknown, but it is related to apparmor3 misbehaving. This did not get detected due to apparmor not logging this in the audit log, and running proftpd unconfined did not change anything. Other permission violations from proftpd did get correctly logged (and fixed in the profile), and thus we did not suspect anything was wrong.

Apparmor 3 being in development combined with us shoehorning it into kernel 4.1 while 4.0 was the latest supported are likely suspects. When we reverted to Apparmor 2 due to seeing other issues, the proftpd problem also disappeared. I thought we had tested it on the stock EL7 kernel as well, but obviously that was not the case.

So, yeah. Always interesting to port things to a new platform, lots of interesting problems to take care of :)
